Issue-967

jkosh44 opened the pull request:
Initial implementation for getsAsync()
Created the following classes:
  AsyncReader
  TransactionImplTest

jkosh44 commented:
This is my first pass at implementing getsAsync, though I haven't had a good chance to look it over and test properly. I started to try and test it but was having trouble creating a transaction to test with. I'm going to give it another go tomorrow.

jkosh44 commented:
Also need to implement close() still

keith-turner commented on modules/core/src/test/java/org/apache/fluo/core/impl/TransactionImplTest.java:
```diff
@@ -0,0 +1,55 @@
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TransactionImplTest {
```
This test should be in a different place. Fluo has Integration test (called ITs) that run against a real Fluo instance (backed by real Accumulo and Zookeeper instances).  These test are at `fluo/modules/integration/src/test/java/org/apache/fluo/integration`.  I would recommend putting your test there under client or impl and extending ITBaseImpl.  After that each test should have a FluoClient called client available.  So should be able to do the following.

```java
@Test
 public void testBasic() {
     try (Transaction tx = client.newTransaction()) {
          tx.set("row1", new Column("col1"), "val1");
          tx.set("row2", new Column("col2"), "val2");
￼          tx.set("row3", new Column("col3"), "val3");
￼ 
          tx.commit();
     }

     try (Transaction tx = client.newTransaction()) {
         //do async gets....
     }
     
}
```

jkosh44 commented:
Just a note, I noticed all of the get methods with Byte were called "get(...)" while all the get methods with Strings were called "gets(...)". I kept this consistent with the async methods.

keith-turner commented on modules/core/src/main/java/org/apache/fluo/core/impl/AsyncReader.java:
```diff
@@ -0,0 +1,100 @@
+    executorService.shutdown();
+  }
+
+  class AsyncGet {
```
Could make this static.  Since it is not static it has an extra hidden reference to `AsyncReader.this` which is not used and makes the objects slightly larger.

keith-turner commented on modules/core/src/main/java/org/apache/fluo/core/impl/AsyncReader.java:
```diff
@@ -0,0 +1,100 @@
+
+  public CompletableFuture<String> gets(String row, Column column, String defaultValue) {
+    Bytes defaultValueBytes = defaultValue == null ? new Bytes() : Bytes.of(defaultValue);
+    return get(Bytes.of(row), column, defaultValueBytes).thenApply(b -> b.toString());
```
Nice, I like the use of thenApply here.

keith-turner commented on modules/core/src/main/java/org/apache/fluo/core/impl/AsyncReader.java:
```diff
@@ -0,0 +1,100 @@
+        }
+      } catch (Exception e) {
+        for (AsyncGet asyncGet : getsList) {
+          asyncGet.res.completeExceptionally(e);
```
If an exception were to occur in the for loop inside the try, then it is possible that some of the futures are already completed.  That case seems unexpected, not sure what if anything we should do about that case.

jkosh44 replied:
How about removing each AsyncGet object from getsList as we complete it?

keith-turner replied:
I was looking at the [javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#completeExceptionally-java.lang.Throwable-) and I don't think anything needs to be done.  The javadoc for completeExceptionally says the following

```
If not already completed, causes invocations of get() and related methods to throw the given exception.
```

So it seems if its already completed, then nothing will happen.

keith-turner commented on modules/core/src/main/java/org/apache/fluo/core/impl/TransactionImpl.java:
```diff
@@ -265,6 +269,42 @@ public Bytes get(Bytes row, Column column) {
+
+  @Override
+  public CompletableFuture<String> getsAsync(String row, Column column) {
+    if (asyncReader == null) {
```
Could shorten these method with a method like the following

```java
private AsyncReader getAsyncReader() {
  if (asyncReader == null) {
￼    asyncReader = new AsyncReader(this);
￼  }
}
```

Then could do the following below

```java
  return getAsyncReader().gets(row, column);
```

keith-turner commented on modules/api/src/main/java/org/apache/fluo/api/client/SnapshotBase.java:
```diff
@@ -165,4 +167,22 @@
+  }
+
+  default CompletableFuture<String> getsAsync(String row, Column column, String defaultValue) {
+    return CompletableFuture.supplyAsync(() -> gets(row, column, defaultValue));
```
Did you see my comment on the issue about supplyAsync?  These request will be put on a JVM wide thread pool and these request do I/O.  So they may block other unrelated work in the JVM, but not completely sure.  I will try to look into this some more.

jkosh44 replied:
I did see but I thought it was OK for the default method to be blocking as long as the implementations weren't blocking. I can remove the default implementation all together.

keith-turner replied:
The [supplyAsync](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#supplyAsync-java.util.function.Supplier-) method will run the task on the ForkJoin common pool which is a JVM wide pool.  I was thinking the default method should just do something like the following to avoid this.

```java
  return CompletableFuture.completedFuture(gets(row, column, defaultValue));
```

keith-turner commented on modules/integration/src/test/java/org/apache/fluo/integration/impl/TransactionImplIT.java:
```diff
@@ -0,0 +1,48 @@
+      Assert.assertEquals("val4", tx.getsAsync("row4", new Column("col4"), "val4").get());
+    } catch (Exception e) {
+      Assert.fail("getsAsync caused the following exception: " + e);
+    }
```
@jkosh44 did you see the comment above?

jkosh44 replied:
Yes, does the testgetAsync() method not cover these cases?

keith-turner replied:
Not quite, the test immediately calls `get()` on the Future and `get()` blocks so its never doing more than one async operation at time. I modified one of the test to do the following and ran it locally, I put in the timing info to see make sure the getAsync calls w/o the get() were fast. 

```java
    try (Transaction tx = client.newTransaction()) {
      long t1 = System.currentTimeMillis();
      Assert.assertEquals("val1", tx.getsAsync("row1", new Column("col1")).get());
      Assert.assertEquals("val2", tx.getsAsync("row2", new Column("col2"), "foo").get());
      Assert.assertEquals("val3", tx.getsAsync("row3", new Column("col3")).get());
      Assert.assertEquals("val4", tx.getsAsync("row4", new Column("col4"), "val4").get());
      long t2 = System.currentTimeMillis();
      System.out.println(t2-t1);
    }

    try (Transaction tx = client.newTransaction()) {
      long t1 = System.currentTimeMillis();
      CompletableFuture<String> val1 = tx.getsAsync("row1", new Column("col1"));
      CompletableFuture<String> val2 = tx.getsAsync("row2", new Column("col2"), "foo");
      CompletableFuture<String> val3 = tx.getsAsync("row3", new Column("col3"));
      CompletableFuture<String> val4 = tx.getsAsync("row4", new Column("col4"), "val4");

      long t2 = System.currentTimeMillis();

      Assert.assertEquals("val1", val1.get());
      Assert.assertEquals("val2", val2.get());
      Assert.assertEquals("val3", val3.get());
      Assert.assertEquals("val4", val4.get());
      long t3 = System.currentTimeMillis();

      System.out.println((t2-t1) + " " + (t3- t2));
    }
```

keith-turner commented:
@jkosh44 I noticed your branch is based off an older commit.  I usually add a git remote called upstream like the following.

```bash
git remote add upstream https://github.com/apache/fluo.git
```
Then with that I added I update my local master with the latest commits from upstream before branching for a new issue.

```bash
# update all remote tracking branches
git fetch --all
git checkout master
# bring local master branch up to date with latest commits
git merge upstream/master
# create new branch from latest master
git checkout -b issue-xyz
```

Don't worry about this PR being based on an older commit, its ok.

jkosh44 commented:
Is it too late to follow the git upstream steps for this issue?

keith-turner commented:
> Is it too late to follow the git upstream steps for this issue?

No, you can try.  If you are new to git I would recommend experimenting in another branch.  Try the following just to see what happens.  Does the rebase command have conflicts?

```
git checkout issue-967
# create a new branch to test rebase
git checkout -b issue-967-rebase-test
git fetch --all
git rebase upstream/master
```

keith-turner commented on modules/core/src/main/java/org/apache/fluo/core/impl/AsyncReader.java:
```diff
@@ -0,0 +1,101 @@
+        for (AsyncGet asyncGet : getsList) {
+          Bytes result = getsMap.get(asyncGet.rc);
+          asyncGet.res.complete(result == null ? defaultValue : result);
+          getsList.remove(asyncGet);
```
The following comments are moot because I don't think things needs to be removed based on earlier comment.  Just sharing some knowledge in following comments.

This could cause a concurrent modification exception.  The foreach has a hidden iterator.  In java, many iterators will throw a concurrent modification exception if the collection is modified during iteration.  Using a while loop with an explicit iterator and calling remove on the iterator is the best way to remove while iterating.  

Also removing items from an arraylist could be an O(N^2) operation.  In the situation of removing from the middle of a list a linked list may be better.

keith-turner commented on modules/core/src/main/java/org/apache/fluo/core/impl/AsyncReader.java:
```diff
@@ -0,0 +1,101 @@
+
+        for (AsyncGet asyncGet : getsList) {
+          Bytes result = getsMap.get(asyncGet.rc);
+          asyncGet.res.complete(result == null ? defaultValue : result);
```
this should be `asyncGet.defaultValue`

jkosh44 commented:
My latest commit should address all the issues mentioned.
I changed to asyncGet.defaultValue.
I no longer remove AsyncGets from getsList after completing them.
I changed the default method to your suggestion.
The rebase command had a conflict at TransactionImpl.java because of one of the imports, I fixed the conflict and then the rebase continued and completed fine

jkosh44 commented:
@keith-turner Also I never commented on the 1.2.0 release. That sounds good to me. If you need someone to work on other Async functionality let me know and I'd be happy to help.

keith-turner commented:
@jkosh44  these changes are looking good.  I am still puzzling out the async story for Fluo.  Currently there are a few areas I have identified and opened issue about.

 * To really get good throughput from async Fluo operations, I feel Accumulo itself needs to be modified to offer async APIs for reading a writing data.  #651 #675
 * For Fluo workers, the number of concurrent transaction could be limited by what will fit into memory instead of the number of threads.  Offering async observers allows Fluo to process as many notifications as will fit into memory instead of limiting it to an arbitrary number of threads as it currently does #968.  This could lead to much higher throughput in conjunction with #651 and #675.
 * For loading data, Fluo should offer async operations #795.  Need to test this by implementing use cases of loading data using Kafka. This could lead to much higher throughput in conjunction with #651 and #675.
 * Internally Fluo uses a quasi asynchronous commit process.  This was written using Guava's Listenable futures.  Now that Fluo is using Java 8 this needs to be converted to CompletableFuture. #722

jkosh44 commented:
I left a comment on #722 and will give it a go. On an unrelated note, I feel like all my commits fail on the continuous-integration Travis CI build. Is there a specific reason? I'm not really sure what I'm doing to cause this.

keith-turner commented:
> On an unrelated note, I feel like all my commits fail on the continuous-integration Travis CI build. Is there a specific reason?

The builds on travis are not always predictable.  Sometimes this is an issue with Travis and sometimes its Fluo.  I have been trying to correct Fluo's flakiness as we run into it.  For your latest commit, the new IT is missing an import stmt for CompletableFuture.  Lets see what happens with Travis after that is fixed.

https://travis-ci.org/apache/fluo/builds/307577351#L2882

keith-turner commented:
I squashed and rebased this and added a few comments in #1099.  Looking at this, it would be nice to have in the next release.

jkosh44 commented:
Just reviewed it and compared it to this one and it looks good to me

keith-turner commented:
This work was merged in #1099
